src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let actionToSave = {};
+  const stateActionsArray = [];
+
+  if (Object.keys(state).length !== 0) {
+    Object.assign(actionToSave, state);
+  }
+
+  for (const action of actions) {
+    for (const value of Object.values(action)) {
+      if (value === 'addProperties') {
+        Object.assign(actionToSave, action.extraData);
+        stateActionsArray.push(Object.assign({}, actionToSave));
+      }
+
+      if (value === 'removeProperties') {
+        for (const key of action.keysToRemove) {
+          delete actionToSave[key];
+        }
+        stateActionsArray.push(Object.assign({}, actionToSave));
+      }
+
+      if (value === 'clear') {
+        stateActionsArray.push({});
+        actionToSave = {};
+      }
+    }
```

Comment:
Good work, let's rewrite it to switch/case statement since we rely on a single action.type value